File: sdk/go/modules/modules.go
// Package modules is the Go SDK surface for the module registry (epic
// memql#4183) -- the runtime inventory of what an instance runs:
// components, integrations, packs, and node-type modules, plus the one
// write, flipping a pack's per-instance enablement.
//
// The package mirrors sdk/go/pack in shape -- a Client wrapping a
// client.Dispatcher, typed methods over SDK-owned values, and errors that
// bubble up. No memqlv1.* import leaks out of the exported surface.
//
// Authorization is server-side: reads answer only to owner/admin callers,
// SetPackEnabled to owners, and a refusal comes back as a *RefusedError
// carrying the server's status code and message. Secret env vars NEVER
// carry a value -- Set is the whole answer for them; that is an engine
// guarantee, not a client courtesy.
package modules

import (
	"context"
	"fmt"

	memqlv1 "github.com/znasllc-io/memql/component/grpc/gen"

	"github.com/znasllc-io/memql/sdk/go/client"
)

// Client exposes the module-registry operations over a Dispatcher. Safe to
// reuse across goroutines once constructed.
type Client struct {
	dispatcher *client.Dispatcher
}

// NewClient wires a modules client to the supplied dispatcher (already
// Run()).
func NewClient(dispatcher *client.Dispatcher) *Client {
	return &Client{dispatcher: dispatcher}
}

// RefusedError is a server-side refusal or failure, carrying the gRPC
// status code the result payload reported.
type RefusedError struct {
	Code    int
	Message string
}

func (e *RefusedError) Error() string {
	return fmt.Sprintf("modules: server refused (code %d): %s", e.Code, e.Message)
}

// Module is one inventory row. Kind is "component" | "integration" |
// "pack" | "node-type"; Scope says which truth tier the state is ("node" =
// the answering binary's registries and env, "cluster" = the shared
// graph).
type Module struct {
	Kind          string
	Name          string
	Description   string
	State         string
	StateDetail   string
	Scope         string
	EnvComponents []string
	FqnPrefixes   []string
	CodeReference string
}

// EnvVar is one manifest-declared environment variable on a module's
// detail surface, evaluated on the answering node. Secret entries carry
// set/unset and nothing else.
type EnvVar struct {
	Name         string
	Description  string
	Secret       bool
	Scope        string
	RequiredFor  []string
	Set          bool
	Value        string
	DefaultValue string
}

// Inventory is a List result: the rows plus which binary answered them.
type Inventory struct {
	Modules           []Module
	ReportingNodeID   string
	ReportingNodeType string
}

// Detail is one module's row plus its env surface.
type Detail struct {
	Module            Module
	EnvVars           []EnvVar
	ReportingNodeID   string
	ReportingNodeType string
}

// PackFlip is the SetPackEnabled outcome. RestartRequired is always true
// in v1 -- the flip changes what each node reads at its next boot, and the
// server says so rather than implying a live toggle.
type PackFlip struct {
	PackDomain      string
	PriorEnabled    bool
	Enabled         bool
	RestartRequired bool
}

func moduleFromProto(m *memqlv1.ModuleInfo) Module {
	return Module{
		Kind:          m.GetKind(),
		Name:          m.GetName(),
		Description:   m.GetDescription(),
		State:         m.GetState(),
		StateDetail:   m.GetStateDetail(),
		Scope:         m.GetScope(),
		EnvComponents: m.GetEnvComponents(),
		FqnPrefixes:   m.GetFqnPrefixes(),
		CodeReference: m.GetCodeReference(),
	}
}

// List returns the full module inventory as the answering node assembles
// it. Owner/admin only.
func (c *Client) List(ctx context.Context) (*Inventory, error) {
	msg := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_ModulesList{
			ModulesList: &memqlv1.ModulesListMsg{},
		},
	}
	resp, err := c.dispatcher.SendAndWait(ctx, msg)
	if err != nil {
		return nil, fmt.Errorf("modules.list: %w", err)
	}
	result := resp.GetModulesListResult()
	if result == nil {
		return nil, fmt.Errorf("modules.list: reply carried no ModulesListResult")
	}
	if result.GetErrorCode() != 0 {
		return nil, &RefusedError{Code: int(result.GetErrorCode()), Message: result.GetErrorMessage()}
	}
	inv := &Inventory{
		ReportingNodeID:   result.GetReportingNodeId(),
		ReportingNodeType: result.GetReportingNodeType(),
	}
	for _, m := range result.GetModules() {
		inv.Modules = append(inv.Modules, moduleFromProto(m))
	}
	return inv, nil
}

// Get returns one module's detail (row + env surface) by (kind, name).
// Owner/admin only; an unknown pair is a *RefusedError with code 5
// (NOT_FOUND).
func (c *Client) Get(ctx context.Context, kind, name string) (*Detail, error) {
	msg := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_ModuleDetail{
			ModuleDetail: &memqlv1.ModuleDetailMsg{Kind: kind, Name: name},
		},
	}
	resp, err := c.dispatcher.SendAndWait(ctx, msg)
	if err != nil {
		return nil, fmt.Errorf("modules.get: %w", err)
	}
	result := resp.GetModuleDetailResult()
	if result == nil {
		return nil, fmt.Errorf("modules.get: reply carried no ModuleDetailResult")
	}
	if result.GetErrorCode() != 0 {
		return nil, &RefusedError{Code: int(result.GetErrorCode()), Message: result.GetErrorMessage()}
	}
	d := &Detail{
		Module:            moduleFromProto(result.GetModule()),
		ReportingNodeID:   result.GetReportingNodeId(),
		ReportingNodeType: result.GetReportingNodeType(),
	}
	for _, v := range result.GetEnvVars() {
		d.EnvVars = append(d.EnvVars, EnvVar{
			Name:         v.GetName(),
			Description:  v.GetDescription(),
			Secret:       v.GetSecret(),
			Scope:        v.GetScope(),
			RequiredFor:  v.GetRequiredFor(),
			Set:          v.GetSet(),
			Value:        v.GetValue(),
			DefaultValue: v.GetDefaultValue(),
		})
	}
	return d, nil
}

// SetPackEnabled flips a pack's per-instance enablement. Owner only;
// audited server-side including refusals. The returned PackFlip's
// RestartRequired says when the flip takes effect (each node's next
// boot, in v1 -- always true).
func (c *Client) SetPackEnabled(ctx context.Context, packDomain string, enabled bool, reason string) (*PackFlip, error) {
	msg := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_SetPackEnabled{
			SetPackEnabled: &memqlv1.SetPackEnabledMsg{
				PackDomain: packDomain,
				Enabled:    enabled,
				Reason:     reason,
			},
		},
	}
	resp, err := c.dispatcher.SendAndWait(ctx, msg)
	if err != nil {
		return nil, fmt.Errorf("modules.setPackEnabled: %w", err)
	}
	result := resp.GetSetPackEnabledResult()
	if result == nil {
		return nil, fmt.Errorf("modules.setPackEnabled: reply carried no SetPackEnabledResult")
	}
	if result.GetErrorCode() != 0 {
		return nil, &RefusedError{Code: int(result.GetErrorCode()), Message: result.GetErrorMessage()}
	}
	return &PackFlip{
		PackDomain:      result.GetPackDomain(),
		PriorEnabled:    result.GetPriorEnabled(),
		Enabled:         result.GetEnabled(),
		RestartRequired: result.GetRestartRequired(),
	}, nil
}

File: sdk/ts/src/client/modules.ts
// ModulesClient is the module-registry surface (epic memql#4183): the
// engine reporting what a cluster RUNS -- components, integrations, packs,
// node-type modules -- and the one registry write, flipping a pack's
// per-instance enablement.
//
// Kept as its own client rather than folded into QueryClient for the same
// reason sdk/go/modules stands alone: the registry is an operator surface
// with its own authorization tier (reads owner/admin, write owner-only),
// not part of the generated query/mutation vocabulary. SDK-owned value
// types below normalize the optional wire fields once, so consumers never
// touch `| undefined` soup; no wire type leaks out of the exported
// surface except through ./wire re-exports the portal deliberately avoids.
//
// Error semantics mirror the proto contract: a result carries
// errorCode/errorMessage INSIDE the payload (0 = OK), and the methods
// here throw on a non-zero code -- the caller sees one failure channel.

import type { Dispatcher } from "./dispatcher.js";
import { newShortId } from "./id.js";
import {
  readServerPayload,
  type ModuleEnvVarWire,
  type ModuleInfoWire,
} from "./wire.js";

export type ModuleKind = "component" | "integration" | "pack" | "node-type";

// Module is one registry row. `scope` says which truth tier the state is:
// "node" (the answering binary's registries and env) or "cluster" (the
// shared graph). stateDetail is the engine's one honest sentence about
// anything non-obvious (e.g. enabled-but-not-loaded pending restart).
export interface Module {
  kind: string;
  name: string;
  description: string;
  state: string;
  stateDetail: string;
  scope: string;
  envComponents: string[];
  fqnPrefixes: string[];
  codeReference: string;
}

// ModuleEnvVar is one manifest-declared environment variable, evaluated on
// the ANSWERING node. A secret entry carries set/unset and nothing else --
// `value` is always "" for secrets, by engine contract, and the SDK does
// not soften that into an optional the UI might "fill in" from elsewhere.
export interface ModuleEnvVar {
  name: string;
  description: string;
  secret: boolean;
  scope: string;
  requiredFor: string[];
  set: boolean;
  value: string;
  defaultValue: string;
}

export interface ModulesInventory {
  modules: Module[];
  reportingNodeId: string;
  reportingNodeType: string;
}

export interface ModuleDetail {
  module: Module;
  envVars: ModuleEnvVar[];
  reportingNodeId: string;
  reportingNodeType: string;
}

// SetPackEnabledOutcome echoes the graph state before and after the flip.
// restartRequired is ALWAYS true in v1 -- a flip changes what each node
// reads at its next boot, never what a running node has loaded -- and the
// UI must render that honestly rather than implying a live toggle.
export interface SetPackEnabledOutcome {
  packDomain: string;
  priorEnabled: boolean;
  enabled: boolean;
  restartRequired: boolean;
}

export interface ModulesCallOptions {
  signal?: AbortSignal;
}

function moduleFromWire(w: ModuleInfoWire | null | undefined): Module {
  return {
    kind: w?.kind ?? "",
    name: w?.name ?? "",
    description: w?.description ?? "",
    state: w?.state ?? "",
    stateDetail: w?.stateDetail ?? "",
    scope: w?.scope ?? "",
    envComponents: w?.envComponents ?? [],
    fqnPrefixes: w?.fqnPrefixes ?? [],
    codeReference: w?.codeReference ?? "",
  };
}

function envVarFromWire(w: ModuleEnvVarWire): ModuleEnvVar {
  return {
    name: w.name ?? "",
    description: w.description ?? "",
    secret: w.secret ?? false,
    scope: w.scope ?? "",
    requiredFor: w.requiredFor ?? [],
    set: w.set ?? false,
    value: w.value ?? "",
    defaultValue: w.defaultValue ?? "",
  };
}

function throwOnPayloadError(
  what: string,
  errorCode: number | undefined,
  errorMessage: string | undefined,
): void {
  if ((errorCode ?? 0) !== 0) {
    throw new Error(`${what}: ${errorMessage || `error code ${errorCode}`}`);
  }
}

export class ModulesClient {
  constructor(private readonly dispatcher: Dispatcher) {}

  async listModules(opts: ModulesCallOptions = {}): Promise<ModulesInventory> {
    const resp = await this.dispatcher.sendAndWait(
      { modulesList: { requestId: newShortId() } },
      opts.signal,
    );
    const payload = readServerPayload(resp);
    if (payload?.kind !== "modulesListResult") {
      throw new Error("list modules: unexpected reply payload");
    }
    throwOnPayloadError("list modules", payload.value.errorCode, payload.value.errorMessage);
    return {
      modules: (payload.value.modules ?? []).map(moduleFromWire),
      reportingNodeId: payload.value.reportingNodeId ?? "",
      reportingNodeType: payload.value.reportingNodeType ?? "",
    };
  }

  async getModuleDetail(
    kind: string,
    name: string,
    opts: ModulesCallOptions = {},
  ): Promise<ModuleDetail> {
    const resp = await this.dispatcher.sendAndWait(
      { moduleDetail: { requestId: newShortId(), kind, name } },
      opts.signal,
    );
    const payload = readServerPayload(resp);
    if (payload?.kind !== "moduleDetailResult") {
      throw new Error("module detail: unexpected reply payload");
    }
    throwOnPayloadError("module detail", payload.value.errorCode, payload.value.errorMessage);
    return {
      module: moduleFromWire(payload.value.module),
      envVars: (payload.value.envVars ?? []).map(envVarFromWire),
      reportingNodeId: payload.value.reportingNodeId ?? "",
      reportingNodeType: payload.value.reportingNodeType ?? "",
    };
  }

  async setPackEnabled(
    packDomain: string,
    enabled: boolean,
    reason: string,
    opts: ModulesCallOptions = {},
  ): Promise<SetPackEnabledOutcome> {
    const resp = await this.dispatcher.sendAndWait(
      { setPackEnabled: { requestId: newShortId(), packDomain, enabled, reason } },
      opts.signal,
    );
    const payload = readServerPayload(resp);
    if (payload?.kind !== "setPackEnabledResult") {
      throw new Error("set pack enabled: unexpected reply payload");
    }
    throwOnPayloadError("set pack enabled", payload.value.errorCode, payload.value.errorMessage);
    return {
      packDomain: payload.value.packDomain ?? packDomain,
      priorEnabled: payload.value.priorEnabled ?? false,
      enabled: payload.value.enabled ?? enabled,
      restartRequired: payload.value.restartRequired ?? true,
    };
  }
}
